src/binutilplusplus.ts

```diff
@@ -0,0 +1,371 @@
+    readonly command: string;
+    readonly findResult: FindBinaryStatus;
+}
+
+export interface ExecFailed {
+    readonly resultKind: 'exec-failed';
+    readonly execProgram: ExternalBinary;
+    readonly command: string;
+}
+
+export interface ExecSucceeded {
+    readonly resultKind: 'exec-succeeded';
+    readonly execProgram: ExternalBinary;
+    readonly command: string;
+    readonly stdout: string;
+}
+
+export interface ParsedExecSucceeded<T> {
+    readonly resultKind: 'exec-succeeded';
+    readonly execProgram: ExternalBinary;
+    readonly command: string;
+    readonly result: T;
+}
+
+export interface ExecErrored {
+    readonly resultKind: 'exec-errored';
+    readonly execProgram: ExternalBinary;
+    readonly command: string;
+    readonly code: number;
+    readonly stderr: string;
+}
+
+
+export interface ExecChildProcessStarted {
+    readonly resultKind: 'exec-child-process-started';
+    readonly execProgram: ExternalBinary;
+    readonly command: string;
+    readonly childProcess: ChildProcess;
+}
+
+export type ParsedExecResult<T> = ExecBinNotFound | ExecFailed | ParsedExecSucceeded<T> | ExecErrored;
+export type ExecResult = ExecBinNotFound | ExecFailed | ExecSucceeded | ExecErrored;
+export type FailedExecResult = ExecBinNotFound | ExecFailed | ExecErrored;
+export type BackgroundExecResult = ExecBinNotFound | ExecChildProcessStarted;
+
+export interface FailureMessageOptions {
+    readonly whatFailed?: string;
+}
+
+export namespace ExecResult {
+    export function tryMap<T>(execResult: ExecResult, fn: (output: string) => T): Errorable<T> {
+        if (execResult.resultKind === 'exec-bin-not-found') {
+            return { succeeded: false, error: [`${execResult.execProgram.displayName} command failed trying to run '${execResult.command}': ${execResult.execProgram.binBaseName} not found`] };
+        }
+        if (execResult.resultKind === 'exec-failed') {
+            return { succeeded: false, error: [`${execResult.execProgram.displayName} command failed trying to run '${execResult.command}': unable to run ${execResult.execProgram.binBaseName}`] };
+        }
+
+        if (execResult.resultKind === 'exec-succeeded') {
+            return { succeeded: true, result: fn(execResult.stdout.trim()) };
+        }
+
+        return { succeeded: false, error: [ execResult.stderr ] };
+    }
+
+    export function map<T>(execResult: ExecResult, fn: (output: string) => T): ParsedExecResult<T> {
+        if (ExecResult.failed(execResult)) {
+            return execResult;
+        }
+
+        return {
+            resultKind: execResult.resultKind,
+            execProgram: execResult.execProgram,
+            command: execResult.command,
+            result: fn(execResult.stdout.trim())
+        };
+    }
+
+    export function failureMessage(execResult: FailedExecResult, options: FailureMessageOptions): string {
+        const err = ExecResult.tryMap(execResult, (s) => s);
+        if (Errorable.failed(err)) {
```

Comment:
I think this check can be removed. We are dealing with `FailedExecResult` so no need to verify if they are failed.

Reply:
We could convert it to an assertion, but the compiler can't prove that passing a `FailedExecResult` to `tryMap` will result in a `Failed` Errorable, so we need _something_ to tell the compiler that it can use the error value of the Errorable.  I could write `as Failed` but the `if` check just seemed a bit more belt and braces.